protocols/perf/src/client/handler.rs

```diff
@@ -0,0 +1,193 @@
+            .push_back(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(ReadyUpgrade::new(crate::PROTOCOL_NAME), ()),
+            })
+    }
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::FullyNegotiatedInbound(FullyNegotiatedInbound {
+                protocol, ..
+            }) => void::unreachable(protocol),
+            ConnectionEvent::FullyNegotiatedOutbound(FullyNegotiatedOutbound {
+                protocol,
+                info: (),
+            }) => {
+                let Command { id, params } = self
+                    .requested_streams
+                    .pop_front()
+                    .expect("opened a stream without a pending command");
+                self.outbound.push(
+                    crate::protocol::send_receive(params, protocol)
+                        .map(move |result| (id, result))
+                        .boxed(),
+                );
+            }
+
+            ConnectionEvent::AddressChange(_)
+            | ConnectionEvent::LocalProtocolsChange(_)
+            | ConnectionEvent::RemoteProtocolsChange(_) => {}
+            ConnectionEvent::DialUpgradeError(DialUpgradeError { info: (), error }) => {
+                let Command { id, .. } = self
+                    .requested_streams
+                    .pop_front()
+                    .expect("requested stream without pending command");
+                self.queued_events
+                    .push_back(ConnectionHandlerEvent::NotifyBehaviour(Event {
+                        id,
+                        result: Err(error.into()),
+                    }));
+            }
+            ConnectionEvent::ListenUpgradeError(ListenUpgradeError { info: (), error }) => {
+                void::unreachable(error)
+            }
+        }
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        self.keep_alive
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            Self::ToBehaviour,
+            Self::Error,
+        >,
+    > {
+        if let Some(event) = self.queued_events.pop_front() {
+            return Poll::Ready(event);
+        }
+
+        if let Poll::Ready(Some((id, result))) = self.outbound.poll_next_unpin(cx) {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Event {
+                id,
+                result: result.map_err(Into::into),
+            }));
+        }
+
+        if self.outbound.is_empty() {
+            self.keep_alive = KeepAlive::No
+        } else {
+            self.keep_alive = KeepAlive::Yes
+        }
```

Comment:
Just compute this inline in `fn connection_keep_alive`.

Reply:
Good idea.